SVM: Add specification document (#643)
anza-xyz/agave: svm/doc/spec.md

# Solana Virtual Machine specification

# Introduction

Several components of the Solana Validator are involved in processing

> of a transaction (or a batch of transactions).  Collectively the

Reviewer: ```suggestion
a transaction (or a batch of transactions).  Collectively, the
```

components responsible for transaction execution are designated as
Solana Virtual Machine (SVM). SVM packaged as a stand-alone library
can be used in applications outside the Solana Validator.

> This document represents specification of the SVM. It covers the API

Reviewer: ```suggestion
This document represents the SVM specification. It covers the API
```

    namely ‘load accounts’ and ‘execute’ stages.

- **SVM Rollups**

    Rollups that need to execute a block but don’t need the other
    components of the validator can benefit from SVM, as it can reduce
    hardware requirements and decentralize the network. This is
    especially useful for Ephemeral Rollups since the cost of compute
    will be higher as a new rollup is created for every user session
    in applications like gaming.

- **SVM Fraud Proofs for Diet Clients**

    A succinct proof of an invalid state transition by the supermajority (SIMD-65)

- **Validator Sidecar for JSON-RPC**

    The RPC needs to be separated from the validator.
    `simulateTransaction` requires replaying the transactions and
    accessing necessary account data.

- **SVM-based Avalanche subnet**

    The SVM would need to be isolated to run within a subnet since the
    consensus and networking functionality would rely on Avalanche
    modules.

- **Modified SVM (SVM+)**

    An SVM type with all the current functionality and extended
    instructions for custom use cases. This would form a superset of
    the current SVM.

# System Context

> In this section SVM is represented as a single entity. We describe its

Reviewer: ```suggestion
In this section, SVM is represented as a single entity. We describe its
```

interfaces to the parts of the Solana Validator external to SVM.

> In the context of Solana Validator the main entity external to SVM is

Reviewer: ```suggestion
In the context of Solana Validator, the main entity external to SVM is
```

bank. It creates an SVM, submits transactions for execution and
receives results of transaction execution from SVM.

![context diagram](/svm/doc/diagrams/context.svg "System Context")

## Interfaces

> In this section we describe the API of using the SVM both in Solana

Reviewer: ```suggestion
In this section, we describe the API of using the SVM both in Solana
```

Validator and in third-party applications.

The interface to SVM is represented by the
`transaction_processor::TransactionBatchProcessor` struct.  To create

> a `TransactionBatchProcessor` object the client need to specify the

Reviewer: ```suggestion
a `TransactionBatchProcessor` object the client needs to specify the
```

`slot`, `epoch`, `epoch_schedule`, `fee_structure`, `runtime_config`,
and `program_cache`.

The main entry point to the SVM is the method

> `load_and_execute_sanitized_transactions`. In addition

Reviewer: ```suggestion
`load_and_execute_sanitized_transactions`. In addition,
```

        - a vector of `Signature`  -- explain which signatures are in this vector
    - `check_results` is a mutable slice of `TransactionCheckResult`
    - `error_counters` is a mutable reference to `TransactionErrorMetrics`
    - `recording_config` is a value of `ExecutionRecordingConfig` configuration parameters
    - `timings` is a mutable reference to `ExecuteTimings`
    - `account_overrides` is an optional reference to `AccountOverrides`
    - `builtin_programs` is an iterator of `Pubkey` that represents builtin programs
    - `log_messages_bytes_limit` is an optional `usize` limit on the size of log messages in bytes
    - `limit_to_load_programs` is a boolean flag that instruct the function to only load the
      programs and do not execute the transactions.

The method returns a value of
`LoadAndExecuteSanitizedTransactionsOutput` which consists of two
vectors
    - a vector of `TransactionLoadResult`, and
    - a vector of `TransactionExecutionResult`.

An integration test `svm_integration` contains an example of
instantiating `TransactionBatchProcessor` and calling its method
`load_and_execute_sanitized_transactions`.

# Functional Model

> In this section we describe the functionality (logic) of the SVM in

Reviewer: ```suggestion
In this section, we describe the functionality (logic) of the SVM in
```

terms of its components, relationships among components, and their
interactions.

On a high level the control flow of SVM consists of loading program
accounts, checking and verifying the loaded accounts, creating
invocation context and invoking RBPF on programs implementing the

> instructions of a transaction. The SVM needs to have access to account
> database, sysvar cache via traits implemented for the corresponding

Reviewer: ```suggestion
instructions of a transaction. The SVM needs to have access to an account
database, and a sysvar cache via traits implemented for the corresponding
```

   - For each `SanitizedTransaction` and `TransactionCheckResult`, we:
        - Calculate the number of signatures in transaction and its cost.
        - Call `load_transaction_accounts`
            - The function is interwined with the struct `CompiledInstruction`
            - Load accounts from accounts DB
            - Extract data from accounts
            - Verify if we've reached the maximum account data size
            - Validate the fee payer and the loaded accounts
            - Validate the programs accounts that have been loaded and checks if they are builtin programs.
            - Return `struct LoadedTransaction` containing the accounts (pubkey and data),
              indices to the excutabe accounts in `TransactionContext` (or `InstructionContext`),
              the transaction rent, and the `struct RentDebit`.
            - Generate a `NonceFull` struct (holds fee subtracted nonce info) when possible, `None` otherwise.
    - Returns `TransactionLoadedResult`, a tuple containing the `LoadTransaction` we obtained from `loaded_transaction_accounts`,
      and a `Option<NonceFull>`.

3. Execute each loaded transactions

>    1. Compute the sum of transaction account balances. This sum is
>       invariant of the transaction execution.

Reviewer: ```suggestion
   1. Compute the sum of transaction accounts' balances. This sum is
      invariant in the transaction execution.
```

            `LoadedPrograms` data structure.
      6. Call `MessageProcessor::process_message` to execute the
      transaction. `MessageProcessor` is contained in
      solana-program-runtime crate. The result of processing message
      is either `ProcessedMessageInfo` which is an i64 wrapped in a
      struct meaning the change in accounts data length, or a
      `TransactionError`, if any of instructions failed to execute
      correctly.
   7. Verify transaction accounts' `RentState` changes (`verify_changes` function)
      - If the account `RentState` pre-transaction processing is rent exempt or unitiliazed, the verification will pass.
      - If the account `RentState` pre-transaction is rent paying:
         - A transition to a state uninitialized or rent exempt post-transaction is not allowed.
         - If its size has changed or its balance has increased, it cannot remain rent paying.
   8. Extract log messages.
   9. Extract inner instructions (`Vec<Vec<InnerInstruction>>`).
   10. Extract `ExecutionRecord` components from transaction context.
   11. Check balances of accounts to match the sum of balances before
       transaction execution.

>    12. Updated loaded transaction accounts to new accounts.

Reviewer: ```suggestion
   12. Update loaded transaction accounts to new accounts.
```

   13. Extract changes in accounts data sizes
   14. Extract return data
   15. Return `TransactionExecutionResult` with wrapping the extracted
       information in `TransactionExecutionDetails`.

4. Prepare the results of loading and executing transactions.

   This includes the following steps for each transactions

>    1. Dump flattened result to info log for an account which pubkey is

Reviewer: ```suggestion
   1. Dump flattened result to info log for an account whose pubkey is
```
